### PR TITLE
Add timeout/retry logic to xcodebuild -list call

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Many users are reporting flakiness (20% failure) in flutter builds on iOS that are caused when the call to xcodebuild -list times out. Until we understand the root cause of how this happens, we can smooth this over with some retry logic.

## Related Issues

https://github.com/flutter/flutter/issues/28415

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes tests for *all* changed/updated/fixed behaviors (See [Test Coverage]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter_tools/lib/src/ios/xcodeproj.dart [hunk 1]
```diff
@@ -161,10 +162,29 @@ class XcodeProjectInterpreter {
...
+    // xcodebuild -list occassionally hangs and we're not sure why.
+    // See: https://github.com/flutter/flutter/issues/28415
+    String out;
+    for (int i = 0; i < 5; i++) {
+      try {
+        final RunResult result = await runCheckedAsync(
+          <String>[_executable, '-list'],
+          workingDirectory: projectPath,
+        ).timeout(const Duration(seconds: 10));
+        // Its okay for this call to fail - it will if the project is
```
Comment (xster) [inline packages/flutter_tools/lib/src/ios/xcodeproj.dart]: It's

**Review by xster — COMMENTED** [review]
LGTM as a temporary measure.

**Comment by Hixie** [comment]
Please don't add timeouts, they are race conditions. (e.g. consider what happens if this command always takes T+1s to run on a particular machine, and you time it out at T seconds every time).

**Comment by jonahwilliams** [comment]
I'm open to alternatives. We have a report of a single command hanging (1 in 5) that I haven't been able to reproduce locally. Should we not do anything?

**Comment by mikhail-tokarev** [comment]
Hi @jonahwilliams,

we tested the PR and unfortunately, it didn't resolve the problem with xcodebuild. With this PR `xcodebuild -list` passed successfully but build stuck on `xcodebuild -project` command :(

```
/usr/local/bin/flutter build ios --debug --no-codesign -v

[  +54 ms] executing: [/Users/greenhouse/programs/flutter/] git rev-parse --abbrev-ref --symbolic @{u}
[  +61 ms] Exit code 128 from: git rev-parse --abbrev-ref --symbolic @{u}
[   +1 ms] fatal: HEAD does not point to a branch
[        ] executing: [/Users/greenhouse/programs/flutter/] git rev-parse --abbrev-ref HEAD
[  +13 ms] Exit code 0 from: git rev-parse --abbrev-ref HEAD
[        ] HEAD
[        ] executing: [/Users/greenhouse/programs/flutter/] git log -n 1 --pretty=format:%H
[  +13 ms] Exit code 0 from: git log -n 1 --pretty=format:%H
[        ] 195364dd0ae72bddf0eb9971c96ae72ea57c80ee
[        ] executing: [/Users/greenhouse/programs/flutter/] git log -n 1 --pretty=format:%ar
[  +16 ms] Exit code 0 from: git log -n 1 --pretty=format:%ar
[        ] 2 days ago
[   +1 ms] executing: [/Users/greenhouse/programs/flutter/] git describe --match v*.*.* --first-parent --long --tags
[  +40 ms] Exit code 0 from: git describe --match v*.*.* --first-parent --long --tags
[        ] v1.3.2-11-g195364dd0
[ +437 ms] Found plugin share at /Users/greenhouse/.pub-cache/hosted/pub.dartlang.org/share-0.5.3/
[ +127 ms] Found plugin share at /Users/greenhouse/.pub-cache/hosted/pub.dartlang.org/share-0.5.3/
[ +280 ms] executing: /usr/bin/defaults read /Users/greenhouse/tmp8y995mm2/ios/Runner/Info CFBundleIdentifier
[  +85 ms] Exit code 0 from: /usr/bin/defaults read /Users/greenhouse/tmp8y995mm2/ios/Runner/Info CFBundleIdentifier
[        ] $(PRODUCT_BUNDLE_IDENTIFIER)
[  +11 ms] Warning: Building for device with codesigning disabled. You will have to manually codesign before deploying to device.
[  +11 ms] Building org.sneath.dadjokes for device (ios)...
[  +18 ms] Removing obsolete reference to flutter_assets from Runner.app
[   +7 ms] Using legacy Xcode build system.
[  +71 ms] executing: [/Users/greenhouse/tmp8y995mm2/ios/] /usr/bin/xcodebuild -list
[+2487 ms] Information about project "Runner":
               Targets:
                   Runner

               Build Configurations:
                   Debug
                   Release

               If no build configuration is specified and -scheme is not passed then "Release" is used.

               Schemes:
                   Runner
                   Runner
[   +6 ms] Trying to resolve native pub services.
[   +2 ms] Looking for YAML at 'pubspec.yaml'
[   +2 ms] No services specified in the manifest
[        ] Found 0 service definition(s).
[   +4 ms] Copying service frameworks to '/Users/greenhouse/tmp8y995mm2/ios/Frameworks'.
[   +2 ms] Creating service definitions manifest at '/Users/greenhouse/tmp8y995mm2/ios/ServiceDefinitions.json'
[  +50 ms] Found plugin share at /Users/greenhouse/.pub-cache/hosted/pub.dartlang.org/share-0.5.3/
[  +18 ms] executing: [/Users/greenhouse/tmp8y995mm2/ios/Runner.xcodeproj/] /usr/bin/xcodebuild -project /Users/greenhouse/tmp8y995mm2/ios/Runner.xcodeproj -target Runner -showBuildSettings
```

**Comment by jonahwilliams** [comment]
@mikhail-tokarev is this reproduce-able locally? If so could you share the output of only `xcodebuild` from `/Users/greenhouse/tmp8y995mm2/` or a similar tmp directory?

**Comment by priitlatt** [comment]
@jonahwilliams Since #28415 doesn't appear to be caused by `xcodebuild -list` and the likes taking very long to complete, but rather that Flutter doesn't detect the output of those commands, we've have come up with another kind of approach to tackle this. Please take a look at PR #29048 - with that change applied, we haven't seen the builds getting stuck any more.

**Comment by jonahwilliams** [comment]
Closing for now

**Pull request closed without merging**